src/storage/src/hummock/iterator/skip_watermark.rs

```diff
@@ -0,0 +1,397 @@
+                                        // current key is not filtered by the watermark at the front.
+                                        #[cfg(debug_assertions)]
+                                        {
+                                            if let Some((next_table_id, next_vnode, _, _)) =
+                                                self.remain_watermarks.front()
+                                            {
+                                                assert!(
+                                                    (next_table_id, next_vnode)
+                                                        > (&key_table_id, &key_vnode)
+                                                );
+                                            }
+                                        }
+                                        return false;
+                                    }
+                                }
+                            }
+                            WatermarkDirection::Descending => {
+                                return match inner_key.cmp(watermark.as_ref()) {
+                                    // Current key as not reached the watermark. Just return.
+                                    Ordering::Less | Ordering::Equal => false,
+                                    // Current key will be filtered by the watermark.
+                                    // Return true to further advance the key.
+                                    Ordering::Greater => true,
+                                };
+                            }
+                        }
+                    }
+                    Ordering::Greater => {
+                        return false;
+                    }
+                }
+            }
+        }
+        false
+    }
+
+    /// Advance the key until iterator invalid or the current key will not be filtered by the latest watermark.
+    /// Calling this method should ensure that the first remaining watermark has been advanced to the current key.
+    ///
+    /// Return a flag indicating whether should later advance the watermark.
+    async fn advance_key(&mut self) -> HummockResult<bool> {
+        if let Some((table_id, vnode, direction, watermark)) = self.remain_watermarks.front() {
+            while self.inner.is_valid() {
+                {
+                    let key = self.inner.key();
+                    let key_table_id = key.user_key.table_id;
+                    let (key_vnode, inner_key) = key.user_key.table_key.split_vnode();
+                    match (&key_table_id, &key_vnode).cmp(&(table_id, vnode)) {
+                        Ordering::Less => {
+                            return Ok(false);
+                        }
+                        Ordering::Equal => {
+                            if direction.filter_by_watermark(inner_key, watermark) {
+                                self.inner.next().await?;
```

Comment:
> https://github.com/risingwavelabs/risingwave/pull/13429#discussion_r1405859699

Currently in our `HummockIterator` trait design, `seek` is more like restarting the iterator instead of advancing to a later keys, and if we use `seek`, we will break this design. For example, some initialization logic of prefetch is triggered in `seek`, and if we keep calling `seek`, we are actually keep restarting a new iterator.

Reply:
discussed offline previously: we are aware that this can be a potential issue but given that using seek will break the degisn and we also [didn't optimize looping over key](https://github.com/risingwavelabs/risingwave/blob/51c76aaa13ac4e81b61d37a07be1ac7afc6631a9/src/storage/src/hummock/iterator/forward_user.rs#L132) with range tombstone, we think it is okay to keep the current logic. Note that with vnode watermark implemented, executor can use it to construct a bounded range for storage iter, which may help to skip the vnode watermarks in advance.